hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/CallerContext.java

```diff
@@ -109,11 +113,25 @@ public String toString() {
 
   /** The caller context builder. */
   public static final class Builder {
-    private final String context;
+    private static final String colon = ":";
+    private final String separator;
+    private final StringBuilder sb = new StringBuilder();
     private byte[] signature;
 
     public Builder(String context) {
-      this.context = context;
+      separator = HADOOP_CALLER_CONTEXT_SEPARATOR_DEFAULT;
+      if (isValid(context)) {
+        sb.append(context);
+      }
+    }
+
+    public Builder(Configuration conf) {
+      separator = conf.get(HADOOP_CALLER_CONTEXT_SEPARATOR_KEY,
+          HADOOP_CALLER_CONTEXT_SEPARATOR_DEFAULT);
+    }
+
+    private boolean isValid(String context) {
+      return context != null && context.length() > 0;
```

Comment:
This is technically not "context" but "item".
Good to have a javadoc explaining what we don't allow.